components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/latest/GitpodPortForwardingService.kt

```diff
@@ -0,0 +1,136 @@
+import io.gitpod.jetbrains.remote.GitpodManager
+import io.gitpod.jetbrains.remote.GitpodPortsService
+import io.gitpod.supervisor.api.Status
+import io.gitpod.supervisor.api.StatusServiceGrpc
+import io.grpc.stub.ClientCallStreamObserver
+import io.grpc.stub.ClientResponseObserver
+import io.ktor.utils.io.*
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.TimeUnit
+
+@Suppress("UnstableApiUsage")
+class GitpodPortForwardingService(private val project: Project) {
+    companion object {
+        const val FORWARDED_PORT_LABEL = "gitpod"
+    }
+
+    private val portsService = service<GitpodPortsService>()
+
+    init { start() }
+
+    private fun start() {
+        if (application.isHeadlessEnvironment) return
+
+        observePortsListWhileProjectIsOpen()
+    }
+
+    private fun observePortsListWhileProjectIsOpen() = application.executeOnPooledThread {
+        while (project.lifetime.status == LifetimeStatus.Alive) {
+            try {
+                observePortsList().get()
+            } catch (throwable: Throwable) {
+                when (throwable) {
+                    is InterruptedException, is CancellationException -> break
+                    else -> thisLogger().error(
+                            "gitpod: Got an error while trying to get ports list from Supervisor. " +
+                                    "Going to try again in a second.",
+                            throwable
+                    )
+                }
+            }
+
+            TimeUnit.SECONDS.sleep(1)
+        }
+    }
+
+    private fun observePortsList(): CompletableFuture<Void> {
+        val completableFuture = CompletableFuture<Void>()
+
+        val statusServiceStub = StatusServiceGrpc.newStub(GitpodManager.supervisorChannel)
+
+        val portsStatusRequest = Status.PortsStatusRequest.newBuilder().setObserve(true).build()
+
+        val portsStatusResponseObserver = object :
+                ClientResponseObserver<Status.PortsStatusRequest, Status.PortsStatusResponse> {
+            override fun beforeStart(request: ClientCallStreamObserver<Status.PortsStatusRequest>) {
+                project.lifetime.onTerminationOrNow { request.cancel("gitpod: Project terminated.", null) }
+            }
+            override fun onNext(response: Status.PortsStatusResponse) { handlePortStatusResponse(response) }
+            override fun onCompleted() { completableFuture.complete(null) }
+            override fun onError(throwable: Throwable) { completableFuture.completeExceptionally(throwable) }
+        }
+
+        statusServiceStub.portsStatus(portsStatusRequest, portsStatusResponseObserver)
+
+        return completableFuture
+    }
+
+    private fun handlePortStatusResponse(response: Status.PortsStatusResponse) {
+        val portNumberToServedStatusMap = mutableMapOf<Int, Boolean>()
+
+        for (port in response.portsList) {
+            val isServed = (port.exposed != null || port.tunneled != null) && port.served
```

Comment:
Why not check their value here but check with `!= null`?

Reply:
If we are going to serve all opened ports, use `port.served` is enough

Reply:
Well observed! Updated! Thank you!